### PR TITLE
Add Event Check-In app & related admin utility actions

### PR DESCRIPTION
**Pull request by jonesnapoleon**
Changes:
- Added `eventcheckin` app & tests, with which events, attendees, and attendances are stored. Checking-in an event works by `POST /api/checkin`-ing attendance token that can be sent as QR codes via admin panel actions.
- Added _migration_ (export registrant objects to attendee objects) admin panel actions on Preevent, Mainevent, and `competition`'s Team.

Why is `CheckInEvent` not implemented as `GenericForeignKey` to Mainevent/Preevent?
- Gala Dinner event doesn't fit into both categories (apps).
- Not worth the time spent for taking care filter rules etc (tested).

### CODE REVIEW

File: arkav/eventcheckin/services.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+logger = logging.getLogger(__name__)
+
+
+class CheckInService():
+
+    @transaction.atomic
+    def checkin(self, checkin_data):
+        attendance = CheckInAttendance.objects.get(token=checkin_data['token'])
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: kalo ga ada gimana? belom dihandle

File: arkav/eventcheckin/services.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            raise ArkavAPIException(
+                detail='Attendee can only check in once',
+                code='attendee_already_checkin',
+                status_code=status.HTTP_400_BAD_REQUEST,
+            )
+
+        try:
+            attendance.checkin_time = timezone.now()
+            attendance.save()
+        except ValueError as e:
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: expected errornya apa? kok kayanya di bagian ini ga bakal ValueError

File: arkav/eventcheckin/services.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            raise ArkavAPIException(
+                detail=str(e),
+                code='checkin_fail',
+                status_code=status.HTTP_400_BAD_REQUEST,
+            )
+
+        return attendance
+
+    @transaction.atomic
+    def migrate_teams(self, teams, events):
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: dipindah ke servicenya competition aja, biar checkinevent ga dependent ke lain2

File: arkav/eventcheckin/services.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            )
+            attendees.append(attendee)
+        for event in events:
+            CheckInAttendance.objects.bulk_create([CheckInAttendance(
+                event=event,
+                attendee=attendee,
+            ) for attendee in attendees])
+
+    @transaction.atomic
+    def migrate_registrants(self, registrants, events):
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: dipindah ke service mainevent ama preevent aja, biar checkinevent ga dependent ke lain2

File: arkav/eventcheckin/services.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        )
+        mail.attach_alternative(mail_html_message, 'text/html')
+        try:
+            django_rq.enqueue(mail.send)
+        except Exception:
+            logger.error('Error mailing {} with subject {}'.format(mail_to, subject))
+
+    def send_token_qr_code(self, attendance):
+        html_template = get_template('token_qr_code_email.html')
+        text_template = get_template('token_qr_code_email.html')
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: dibikin file txt nya aja

isinya either link ke gambar qr code, atau tulis kalo qrcode tidak muncul karena harus dilihat di html

File: arkav/eventcheckin/tests.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        data = {
+            'token': self.attendance_unattended.token,
+        }
+        res = self.client.post(url, data=data, format='json')
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+
+        self.attendance_unattended.refresh_from_db()
+        self.assertTrue(self.attendance_unattended.is_checked_in)
+
+    def test_checkin_already_attended(self):
```
Comment (yonasadiel) [inline arkav/eventcheckin/tests.py]: tambah test buat token salah

File: arkav/eventcheckin/views.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+from arkav.eventcheckin.serializers import CheckInRequestSerializer
+from arkav.eventcheckin.serializers import CheckInResponseSerializer
+from arkav.eventcheckin.services import CheckInService
+from arkav.utils.exceptions import ArkavAPIException
+from rest_framework import generics
+from rest_framework import status
+from django.utils.decorators import method_decorator
+from django.views.decorators.csrf import ensure_csrf_cookie
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework.response import Response
```
Comment (yonasadiel) [inline arkav/eventcheckin/views.py]: importnya tolong diurutin path dong

File: arkav/eventcheckin/migrations/0001_initial.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+# Generated by Django 2.2.8 on 2020-01-20 18:14
```
Comment (yonasadiel) [inline arkav/eventcheckin/migrations/0001_initial.py]: makemigrations lagi kalo sudah nambah last_sent_token ya

File: arkav/eventcheckin/models.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+class CheckInAttendance(models.Model):
+    """
+    Many-to-many through/pivot table of CheckInEvent to CheckInAttendee
+    """
+    event = models.ForeignKey(to=CheckInEvent, related_name='event_attendee', on_delete=models.CASCADE)
+    attendee = models.ForeignKey(to=CheckInAttendee, related_name='event_attendee', on_delete=models.CASCADE)
+
+    token = models.UUIDField(unique=True, default=uuid.uuid4, editable=False)
+    checkin_time = models.DateTimeField(auto_now_add=False, null=True, blank=True)
```
Comment (yonasadiel) [inline arkav/eventcheckin/models.py]: tambahin last_sent_token buat track kapan kode dikirim, biar tau mana yg ga kekirim

File: arkav/eventcheckin/services.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+        context = {
+            'event': attendance.event,
+            'attendee': attendance.attendee,
+            'token': attendance.token,
+        }
+
+        subject = '[Arkavidia] {} - Check-In QR Code'.format(attendance.event)
+        self.send_email(attendance.attendee, subject, context, text_template, html_template)
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: ini masukin django_rq

File: arkav/eventcheckin/services.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        mail_to = attendee.email
+
+        mail = EmailMultiAlternatives(
+            subject=subject,
+            body=mail_text_message,
+            to=[mail_to],
+        )
+        mail.attach_alternative(mail_html_message, 'text/html')
+        try:
+            django_rq.enqueue(mail.send)
```
Comment (yonasadiel) [inline arkav/eventcheckin/services.py]: ini ga usah dienqueue

File: arkav/eventcheckin/services.py (hunk 2 above)
Comment (hashshura) [reply arkav/eventcheckin/services.py]: Gaakan ada error sih keknya, kelewat review kemarin. Kuhapus ae ya

**Pull request merged**